server/app/services/program/predicate/PredicateGenerator.java

```diff
@@ -0,0 +1,290 @@
+      DynamicForm predicateForm, ReadOnlyQuestionService roQuestionService)
+      throws QuestionNotFoundException {
+    Multimap<Integer, LeafOperationExpressionNode> leafNodes = LinkedHashMultimap.create();
+
+    HashSet<String> consumedMultiValueKeys = new HashSet<>();
+    final PredicateAction predicateAction;
+
+    try {
+      predicateAction = PredicateAction.valueOf(predicateForm.get("predicateAction"));
+    } catch (IllegalArgumentException e) {
+      throw new BadRequestException(
+          String.format(
+              "Missing or unknown predicateAction: %s", predicateForm.get("predicateAction")));
+    }
+
+    for (String key : predicateForm.rawData().keySet()) {
+      Matcher singleValueMatcher = SINGLE_PREDICATE_VALUE_FORM_KEY_PATTERN.matcher(key);
+      Matcher multiValueMatcher = MULTI_PREDICATE_VALUE_FORM_KEY_PATTERN.matcher(key);
+
+      int groupId;
+      long questionId;
+      Scalar scalar;
+      Operator operator;
+      PredicateValue predicateValue;
+
+      if (singleValueMatcher.find()) {
+        Pair<Integer, Long> groupIdQuestionIdPair =
+            getGroupIdAndQuestionId(roQuestionService, singleValueMatcher);
+        groupId = groupIdQuestionIdPair.getLeft();
+        questionId = groupIdQuestionIdPair.getRight();
+        Pair<Scalar, Operator> scalarOperatorPair = getScalarAndOperator(predicateForm, questionId);
+        scalar = scalarOperatorPair.getLeft();
+        operator = scalarOperatorPair.getRight();
+
+        predicateValue =
+            parsePredicateValue(scalar, operator, predicateForm.get(key), ImmutableList.of());
+      } else if (multiValueMatcher.find() && !consumedMultiValueKeys.contains(key)) {
+        Pair<Integer, Long> groupIdQuestionIdPair =
+            getGroupIdAndQuestionId(roQuestionService, multiValueMatcher);
+        groupId = groupIdQuestionIdPair.getLeft();
+        questionId = groupIdQuestionIdPair.getRight();
+        Pair<Scalar, Operator> scalarOperatorPair = getScalarAndOperator(predicateForm, questionId);
+        scalar = scalarOperatorPair.getLeft();
+        operator = scalarOperatorPair.getRight();
+
+        ImmutableList<String> multiSelectKeys =
+            predicateForm.rawData().keySet().stream()
+                .filter(
+                    filteredKey ->
+                        filteredKey.startsWith(
+                            String.format(
+                                "group-%d-question-%d-predicateValues", groupId, questionId)))
+                .collect(ImmutableList.toImmutableList());
+
+        consumedMultiValueKeys.addAll(multiSelectKeys);
+
+        ImmutableList<String> rawPredicateValues =
+            multiSelectKeys.stream()
+                .map(predicateForm.rawData()::get)
+                .collect(ImmutableList.toImmutableList());
+
+        predicateValue = parsePredicateValue(scalar, operator, "", rawPredicateValues);
+      } else {
+        continue;
+      }
+
+      leafNodes.put(
```

Comment:
optional. FWIW, I believe the code above and below here are effectively separate with the above making leafNodes and the below consuming it, you could break that out into two functions just to have some organization.

Reply:
done